infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java

```diff
@@ -0,0 +1,396 @@
+import org.hibernate.ogm.model.spi.AssociationOperationType;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.infinispan.client.hotrod.MetadataValue;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.client.hotrod.VersionedValue;
+import org.infinispan.commons.util.CloseableIterator;
+import org.infinispan.query.dsl.FilterConditionContext;
+import org.infinispan.query.dsl.Query;
+import org.infinispan.query.dsl.QueryBuilder;
+import org.infinispan.query.dsl.QueryFactory;
+
+/**
+ *  Some implementation notes for evolution:
+ *
+ *  - QueryableGridDialect can't be implemented as "native queries" in Hot Rod are DSL based
+ *    and have no String representation; this might change as Hot Rod exposes the underlying
+ *    query representation which is similar to HQL; alternatively we could look at exposing
+ *    native queries in some way other than a String.
+ *
+ *  - OptimisticLockingAwareGridDialect can't be implemented as it requires an atomic replace
+ *    operation on a subset of the columns, while atomic operations in Hot Rod have to involve
+ *    either the full value or the version metadata of Infinispan's VersionedValue.
+ *
+ *  - IdentityColumnAwareGridDialect can't work out of the box. I suspect we could do this but
+ *    would need extending the Infinispan server deployment with some extension such as a
+ *    custom script to be invoked from the client.
+ *
+ *  - BatchableGridDialect could probably be implemented.
+ *
+ * @author Sanne Grinovero
+ */
+public class InfinispanRemoteDialect<EK,AK,ISK> extends BaseGridDialect implements MultigetGridDialect {
+
+	private static final Log log = LoggerFactory.getLogger();
+
+	private final InfinispanRemoteDatastoreProvider provider;
+
+	public InfinispanRemoteDialect(InfinispanRemoteDatastoreProvider provider) {
+		this.provider = Objects.requireNonNull( provider );
+	}
+
+	@Override
+	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+		final String cacheName = key.getTable();
+		ProtoStreamMappingAdapter mapper = provider.getDataMapperForCache( cacheName );
+		ProtostreamId idBuffer = mapper.createIdPayload( key.getColumnNames(), key.getColumnValues() );
+		VersionedValue<ProtostreamPayload> v = mapper.withinCacheEncodingContext( c -> c.getVersioned( idBuffer ) );
+		if ( v == null ) {
+			return null;
+		}
+		ProtostreamPayload payload = v.getValue();
+		if ( payload == null ) {
+			return null;
+		}
+		long version = v.getVersion();
+		VersionedTuple versionedTuple = payload.toVersionedTuple();
+		versionedTuple.setVersion( version );
+		return versionedTuple;
+	}
+
+	@Override
+	public Tuple createTuple(EntityKey key, TupleContext tupleContext) {
+		return new VersionedTuple( true );
+	}
+
+	@Override
+	public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext) {
+		VersionedTuple versionedTuple = (VersionedTuple) tuple;
+		final long knownVersion = versionedTuple.getVersion();
```

Comment:
thanks. It's needed by the block of code which is commented out below.. I'll remove all commented code and this.